[caffe2] Remove IValue include from operator.h

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #67096
* __->__ #67094

ivalue.h includes Tensor.h, so creating a compilation barrier between
operator.h and ivalue.h means non-exported caffe2 ops don't need to be
rebuilt when developing PyTorch.

Differential Revision: [D32289812](https://our.internmc.facebook.com/intern/diff/D32289812)

peterbell10 committed:
[caffe2] Remove IValue include from operator.h

ivalue.h includes Tensor.h, so creating a compilation barrier between
operator.h and ivalue.h means non-exported caffe2 ops don't need to be
rebuilt as often when developing PyTorch.

[ghstack-poisoned]

peterbell10 committed:
Update on "[caffe2] Remove IValue include from operator.h"


ivalue.h includes Tensor.h, so creating a compilation barrier between
operator.h and ivalue.h means non-exported caffe2 ops don't need to be
rebuilt as often when developing PyTorch.

[ghstack-poisoned]

peterbell10 committed:
Update on "[caffe2] Remove IValue include from operator.h"


ivalue.h includes Tensor.h, so creating a compilation barrier between
operator.h and ivalue.h means non-exported caffe2 ops don't need to be
rebuilt as often when developing PyTorch.

[ghstack-poisoned]

peterbell10 committed:
Update on "[caffe2] Remove IValue include from operator.h"


ivalue.h includes Tensor.h, so creating a compilation barrier between
operator.h and ivalue.h means non-exported caffe2 ops don't need to be
rebuilt as often when developing PyTorch.

[ghstack-poisoned]

peterbell10 committed:
Update on "[caffe2] Remove IValue include from operator.h"


ivalue.h includes Tensor.h, so creating a compilation barrier between
operator.h and ivalue.h means non-exported caffe2 ops don't need to be
rebuilt when developing PyTorch.

[ghstack-poisoned]

peterbell10 committed:
Update on "[caffe2] Remove IValue include from operator.h"


ivalue.h includes Tensor.h, so creating a compilation barrier between
operator.h and ivalue.h means non-exported caffe2 ops don't need to be
rebuilt when developing PyTorch.

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32289812).

dagitses commented on caffe2/core/IValueInterface.h:
```diff
@@ -0,0 +1,51 @@
+  }
+
+  const c10::IValue& at(size_t idx) const;
+  std::vector<caffe2::Tensor> toTensorVector(size_t idx) const;
```
please document the preconditions of these functions, e.g.
`// Expects(isTensorList(idx))` for `toTensorVector`, presumably

peterbell10 commented on caffe2/core/IValueInterface.cc:
```diff
@@ -0,0 +1,179 @@
+  if (values_[0].isTensorList()) {
+    // if the first input is a tensor list, we get input tensors by indexing
+    // into that list. currently, this means that only tensors from that list
+    // are accessible as inputs. any hypothetical input tensors that come after
```
The logic here comes straight from the existing translation code,  so I presume it's needed. Caffe2 also isn't being actively developed so I wouldn't feel comfortable changing any core functionality.

dagitses replied:
sounds good to me.

peterbell10 commented on caffe2/core/IValueInterface.cc:
```diff
@@ -0,0 +1,179 @@
+  size_(values.size()) {
+  auto values_tmp = std::make_unique<c10::IValue[]>(size_);
+  std::copy_n(values.data(), size_, values_tmp.get());
+  values_ = values_tmp.release();
```
I gave it a try and you can't even call `size()` on a vector of incomplete type. I just don't think `std::vector` was intended to support incomplete types.

dagitses replied:
OK how about defining `values_` as `std::unique_ptr<c10::IValue[]>`?

peterbell10 replied:
I get the following error:
```
/usr/include/c++/9/bits/unique_ptr.h: In instantiation of ‘typename std::enable_if<std::is_convertible<_Up (*)[], _Tp (*)[]>::value>::type std::default_delete<_Tp []>::operator()(_Up*) const [with _Up = c10::IValue; _Tp = c10::IValue; typename std::enable_if<std::is_convertible<_Up (*)[], _Tp (*)[]>::value>::type = void]’:
/usr/include/c++/9/bits/unique_ptr.h:559:17:   required from ‘std::unique_ptr<_Tp [], _Dp>::~unique_ptr() [with _Tp = c10::IValue; _Dp = std::default_delete<c10::IValue []>]’
/home/peter/git/pytorch/caffe2/core/IValueInterface.h:24:47:   required from here
/usr/include/c++/9/bits/unique_ptr.h:113:16: error: invalid application of ‘sizeof’ to incomplete type ‘c10::IValue’
  113 |  static_assert(sizeof(_Tp)>0,
```

peterbell10 commented on caffe2/core/IValueInterface.h:
```diff
@@ -0,0 +1,51 @@
+  std::vector<T> toVec(size_t idx) const;
+
+private:
+  c10::IValue *values_;
```
You mean why is it an array and not just a single `IValue`?  It means we only need one memory allocation, and allows things like `compute_input_size` to happen in one function without repeatedly passing in and out of the interface.

peterbell10 committed:
Update on "[caffe2] Remove IValue include from operator.h"


ivalue.h includes Tensor.h, so creating a compilation barrier between
operator.h and ivalue.h means non-exported caffe2 ops don't need to be
rebuilt when developing PyTorch.

Differential Revision: [D32289812](https://our.internmc.facebook.com/intern/diff/D32289812)

[ghstack-poisoned]

peterbell10 committed:
Update on "[caffe2] Remove IValue include from operator.h"


ivalue.h includes Tensor.h, so creating a compilation barrier between
operator.h and ivalue.h means non-exported caffe2 ops don't need to be
rebuilt when developing PyTorch.

Differential Revision: [D32289812](https://our.internmc.facebook.com/intern/diff/D32289812)

[ghstack-poisoned]

peterbell10 commented:
@dagitses PTAL

peterbell10 committed:
Update on "[caffe2] Remove IValue include from operator.h"


ivalue.h includes Tensor.h, so creating a compilation barrier between
operator.h and ivalue.h means non-exported caffe2 ops don't need to be
rebuilt when developing PyTorch.

Differential Revision: [D32289812](https://our.internmc.facebook.com/intern/diff/D32289812)

[ghstack-poisoned]

dagitses commented on caffe2/core/IValueInterface.h:
```diff
@@ -0,0 +1,51 @@
+
+TORCH_API caffe2::Tensor contiguous(caffe2::Tensor);
+
+class TORCH_API IValueInterface {
```
reminder about this comment.

peterbell10 replied:
https://github.com/pytorch/pytorch/blob/e701cd6ebc11aa9171d32bd1626e938b165f60c9/caffe2/core/IValueInterface.h#L16-L20

dagitses replied:
thanks. sorry was looking at wrong commit.

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32289812).

peterbell10 committed:
Update on "[caffe2] Remove IValue include from operator.h"


ivalue.h includes Tensor.h, so creating a compilation barrier between
operator.h and ivalue.h means non-exported caffe2 ops don't need to be
rebuilt when developing PyTorch.

Differential Revision: [D32289812](https://our.internmc.facebook.com/intern/diff/D32289812)

[ghstack-poisoned]

peterbell10 committed:
Update on "[caffe2] Remove IValue include from operator.h"


ivalue.h includes Tensor.h, so creating a compilation barrier between
operator.h and ivalue.h means non-exported caffe2 ops don't need to be
rebuilt when developing PyTorch.

Differential Revision: [D32289812](https://our.internmc.facebook.com/intern/diff/D32289812)

[ghstack-poisoned]

peterbell10 committed:
Update on "[caffe2] Remove IValue include from operator.h"


ivalue.h includes Tensor.h, so creating a compilation barrier between
operator.h and ivalue.h means non-exported caffe2 ops don't need to be
rebuilt when developing PyTorch.

Differential Revision: [D32289812](https://our.internmc.facebook.com/intern/diff/D32289812)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32289812).

dagitses commented on caffe2/core/IValueInterface.h:
```diff
@@ -0,0 +1,62 @@
+class TORCH_API IValueInterface {
+public:
+  IValueInterface(): values_(nullptr), size_(0) {}
+  IValueInterface(c10::ArrayRef<c10::IValue> values);
```
are you able to mark this explicit?

peterbell10 replied:
done

peterbell10 committed:
Update on "[caffe2] Remove IValue include from operator.h"


ivalue.h includes Tensor.h, so creating a compilation barrier between
operator.h and ivalue.h means non-exported caffe2 ops don't need to be
rebuilt when developing PyTorch.

Differential Revision: [D32289812](https://our.internmc.facebook.com/intern/diff/D32289812)

[ghstack-poisoned]

peterbell10 committed:
Update on "[caffe2] Remove IValue include from operator.h"


ivalue.h includes Tensor.h, so creating a compilation barrier between
operator.h and ivalue.h means non-exported caffe2 ops don't need to be
rebuilt when developing PyTorch.

Differential Revision: [D32289812](https://our.internmc.facebook.com/intern/diff/D32289812)

[ghstack-poisoned]

peterbell10 committed:
Rebase after master CI break on "[caffe2] Remove IValue include from operator.h"


ivalue.h includes Tensor.h, so creating a compilation barrier between
operator.h and ivalue.h means non-exported caffe2 ops don't need to be
rebuilt when developing PyTorch.

Differential Revision: [D32289812](https://our.internmc.facebook.com/intern/diff/D32289812)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32289812).

peterbell10 committed:
Only build IValueInterface when necessary on "[caffe2] Remove IValue include from operator.h"


ivalue.h includes Tensor.h, so creating a compilation barrier between
operator.h and ivalue.h means non-exported caffe2 ops don't need to be
rebuilt when developing PyTorch.

Differential Revision: [D32289812](https://our.internmc.facebook.com/intern/diff/D32289812)

[ghstack-poisoned]

peterbell10 committed:
Rebase and fix merge conflicts on "[caffe2] Remove IValue include from operator.h"


ivalue.h includes Tensor.h, so creating a compilation barrier between
operator.h and ivalue.h means non-exported caffe2 ops don't need to be
rebuilt when developing PyTorch.

Differential Revision: [D32289812](https://our.internmc.facebook.com/intern/diff/D32289812)

[ghstack-poisoned]

peterbell10 committed:
Rebase on "[caffe2] Remove IValue include from operator.h"


ivalue.h includes Tensor.h, so creating a compilation barrier between
operator.h and ivalue.h means non-exported caffe2 ops don't need to be
rebuilt when developing PyTorch.

Differential Revision: [D32289812](https://our.internmc.facebook.com/intern/diff/D32289812)

[ghstack-poisoned]

peterbell10 committed:
Update on "[caffe2] Remove IValue include from operator.h"


ivalue.h includes Tensor.h, so creating a compilation barrier between
operator.h and ivalue.h means non-exported caffe2 ops don't need to be
rebuilt when developing PyTorch.

Differential Revision: [D32289812](https://our.internmc.facebook.com/intern/diff/D32289812)

[ghstack-poisoned]

peterbell10 committed:
Rebase on "[caffe2] Remove IValue include from operator.h"


ivalue.h includes Tensor.h, so creating a compilation barrier between
operator.h and ivalue.h means non-exported caffe2 ops don't need to be
rebuilt when developing PyTorch.

Differential Revision: [D32289812](https://our.internmc.facebook.com/intern/diff/D32289812)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32289812).

peterbell10 committed:
Update on "[caffe2] Remove IValue include from operator.h"


ivalue.h includes Tensor.h, so creating a compilation barrier between
operator.h and ivalue.h means non-exported caffe2 ops don't need to be
rebuilt when developing PyTorch.

Differential Revision: [D32289812](https://our.internmc.facebook.com/intern/diff/D32289812)

[ghstack-poisoned]

peterbell10 committed:
Update on "[caffe2] Remove IValue include from operator.h"


ivalue.h includes Tensor.h, so creating a compilation barrier between
operator.h and ivalue.h means non-exported caffe2 ops don't need to be
rebuilt when developing PyTorch.

Differential Revision: [D32289812](https://our.internmc.facebook.com/intern/diff/D32289812)

[ghstack-poisoned]
